Realm/Realm/Handles/SessionHandle.cs

```diff
@@ -224,14 +249,88 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
+                        syncConfig.OnSessionError?.Invoke(session, exception);
+                    }
+                }
+                else
+                {
+                    // after deprecation: this will need to go when Session.Error is fully deprecated
+                    Session.RaiseError(session, exception);
+                }
             }
             catch (Exception ex)
             {
                 Logger.Default.Log(LogLevel.Warn, $"An error has occurred while handling a session error: {ex}");
             }
         }
 
+        [MonoPInvokeCallback(typeof(NativeMethods.NotifyBeforeClientReset))]
+        private static bool NotifyBeforeClientReset(IntPtr beforeFrozen, IntPtr managedSyncConfigurationHandle)
+        {
+            try
+            {
+                var syncConfigHandle = GCHandle.FromIntPtr(managedSyncConfigurationHandle);
+                var syncConfig = (SyncConfigurationBase)syncConfigHandle.Target;
+
+                if (syncConfig.ClientResetHandler == null)
+                {
+                    return true;
+                }
+
+                var discardLocalResetHandler = (DiscardLocalResetHandler)syncConfig.ClientResetHandler;
+                var schema = syncConfig.GetSchema();
+
+                if (discardLocalResetHandler.OnBeforeReset != null)
+                {
+                    using var realmBefore = new Realm(new UnownedRealmHandle(beforeFrozen), syncConfig, schema);
+                    discardLocalResetHandler.OnBeforeReset(realmBefore);
+                }
+            }
+            catch (Exception ex)
+            {
+                Logger.Default.Log(LogLevel.Error, $"An error has occurred while executing SyncConfigurationBase.OnBeforeReset during a client reset: {ex}");
+                return false;
+            }
+
+            return true;
+        }
+
+        [MonoPInvokeCallback(typeof(NativeMethods.NotifyAfterClientReset))]
+        private static bool NotifyAfterClientReset(IntPtr beforeFrozen, IntPtr after, IntPtr managedSyncConfigurationHandle)
+        {
+            try
+            {
+                var syncConfigHandle = GCHandle.FromIntPtr(managedSyncConfigurationHandle);
+                var syncConfig = (SyncConfigurationBase)syncConfigHandle.Target;
+
+                if (syncConfig.ClientResetHandler == null)
+                {
+                    return true;
+                }
+
+                var discardLocalResetHandler = (DiscardLocalResetHandler)syncConfig.ClientResetHandler;
+                var schema = syncConfig.GetSchema();
+
+                if (discardLocalResetHandler.OnAfterReset != null)
+                {
+                    using var realmBefore = new Realm(new UnownedRealmHandle(beforeFrozen), syncConfig, schema);
+                    using var realmAfter = new Realm(new UnownedRealmHandle(after), syncConfig, schema);
+                    discardLocalResetHandler.OnAfterReset(realmBefore, realmAfter);
+                }
```

Comment:
We could rework this slightly like (same thing can be done for BeforeReset).

```suggestion
                if (syncConfig.ClientResetHandler is DiscardLocalResetHandler discardLocalHandler &&
                    discardLocalHandler.OnAfterReset != null)
                {
                    var schema = syncConfig.GetSchema();
                    using var realmBefore = new Realm(new UnownedRealmHandle(beforeFrozen), syncConfig, schema);
                    using var realmAfter = new Realm(new UnownedRealmHandle(after), syncConfig, schema);
                    discardLocalHandler.OnAfterReset(realmBefore, realmAfter);
                }
```